<h2 id="요구사항">요구사항</h2>
<p>👉 매니저 등록 요청 시 로그를 남기고 싶어요!
<code>@Transactional</code>의 옵션 중 하나를 활용하여 매니저 등록과 로그 기록이 각각 독립적으로 처리될 수 있도록 해봅시다.</p>
<ul>
<li>매니저 등록 요청을 기록하는 로그 테이블을 만들어주세요.<ul>
<li>DB 테이블명: <code>log</code></li>
</ul>
</li>
<li>매니저 등록과는 별개로 로그 테이블에는 항상 요청 로그가 남아야 해요.<ul>
<li>매니저 등록은 실패할 수 있지만, 로그는 반드시 저장되어야 합니다.</li>
<li>로그 생성 시간은 반드시 필요합니다.</li>
<li>그 외 로그에 들어가는 내용은 원하는 정보를 자유롭게 넣어주세요.</li>
</ul>
</li>
</ul>
<hr />
<pre><code class="language-java">@Getter
@Entity
@NoArgsConstructor
@Table(name = &quot;logs&quot;)
public class Log {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY) // 일정 id
    @JoinColumn(name = &quot;manager_id&quot;, nullable = false)
    private Manager manager;

    @Column(nullable = false)
    private Boolean isFailed;

    @Column(nullable = false)
    private LocalDateTime time;

    public Log(Manager manager, Boolean isFailed) {
        this.manager = manager;
        this.isFailed = isFailed;
        this.time = LocalDateTime.now();
    }
}</code></pre>
<p>매니저 등록 성공/실패 로그 저장하는 테이블 엔티티
나는 요구사항에 추가해서 <code>isFailed</code> -&gt; 실패여부 Boolean 필드도 추가해줬다</p>
<p>이 때문에 좀 생각할게 많았다</p>
<blockquote>
<p>하지만 TIL을 쓸 수 있게 되지^^</p>
</blockquote>
<hr />
<pre><code class="language-java">        try {

            Manager savedManagerUser = managerRepository.save(newManagerUser);

            logService.saveLog(savedManagerUser,false);

            return new ManagerSaveResponse(
                savedManagerUser.getId(),
                new UserResponse(managerUser.getId(), managerUser.getEmail())
            );

        } catch (Exception e) {

            logService.saveLog(newManagerUser, true);

            throw new ManagerException(&quot;매니저 등록 실패했습니다.&quot;);

        }</code></pre>
<p>우선 나는 매니저 패키지와 로그 패키지를 도메인으로 나누어서 개발했다
그리고 </p>
<pre><code class="language-java">    @Transactional
    public ManagerSaveResponse saveManager(CustomUserPrincipal userPrincipal, long todoId, ManagerSaveRequest managerSaveRequest) {

    @Transactional(propagation = Propagation.REQUIRES_NEW)
    public void saveLog(Manager manager, boolean isFailed) {
</code></pre>
<p>이처럼 manager save Service에는 일반 Transactional
log save Service는 REQUIRES_NEW를 사용해서 상위 트랜잭션에 포함되지 않고 독립적인 트랜잭션을 탈 수 있도록 했다</p>
<blockquote>
<p>로그 생성은 매니저 생성과 다르게 항상 실패하든, 성공하든 항상 저장되어야 하기에 독립적인 트랜잭션을 부여함</p>
</blockquote>
<hr />
<pre><code class="language-java">        try {

            Manager savedManagerUser = managerRepository.save(newManagerUser);

            logService.saveLog(savedManagerUser,false);

            return new ManagerSaveResponse(
                savedManagerUser.getId(),
                new UserResponse(managerUser.getId(), managerUser.getEmail())
            );

        } catch (Exception e) {

            logService.saveLog(newManagerUser, true);

            throw new ManagerException(&quot;매니저 등록 실패했습니다.&quot;);

        }</code></pre>
<p>그리고 manager Service에선 트라이캐치로 감싸줬다 </p>
<blockquote>
<p>매니저 저장 실패 시 다른 로직을 수행해야 하기 때문에</p>
</blockquote>
<ul>
<li>Log의 isFailed 값에 true를 적용해줘야함 성공할 땐 false고</li>
</ul>
<br />

<p>그래서 트라이 캐치로 감쌌는데 고려해야할 부분이 몇 가지 있었음</p>
<br />

<p>🤔 예외를 던져야 하는데 어떤걸 던져야 하지?</p>
<ul>
<li>보통 여태까지는 트라이 캐치문 없이 CustomException을 날렸는데
catch에서 어떤 예외를 캐치해줘야할지 모르겠었음</li>
</ul>
<blockquote>
<p>예외는 가장 큰 단위로 잡는다. 무슨 예외가 터질 지 모르니께
-&gt; 그리고 마지막에 catch 마지막에 CustomException을 던져줘서 롤백 시킨다
-&gt; 왜냐면 트라이 캐치로 예외 터지는 부분을 감싸면 예외가 터져도 트라이 캐치때문에 롤백이 안됨. 그래서 예외를 던져줘야함</p>
</blockquote>
<p>정리</p>
<ol>
<li>REQUIRES_NEW로 트랜잭션 분리함</li>
</ol>
<p>-&gt; 매니저 등록 실패하더라도 로그 생성은 실패하지 않음 둘은 다른 트랜잭션이니까
2. try-catch로 매니저 등록 성공/실패에 따라 다른 로그(isFailed)생성
3. throw new로 Custom 예외 날려서 사용자에게 알리기</p>
<br />

<p>‼️아직 고려해야할 점이 남아있음</p>
<p>현재 catch하는 예외는 Exception 임
요구사항에 특정 예외를 명시하지 않고
어떤 이유로든 매니저 등록에 실패해도 로그를 남겨야하기에 Exception 사용
-&gt; 구체적인 예외를 처리해야한다면 어떻게 처리하는게 좋을까?
-&gt; 예를 들어 나는 매니저 정보 내 투두 정보가 잘못돼서 예외를 날리고 싶은데, 실제 날려지는 예외는 유저 정보가 잘못 된거라면?
-&gt; 근데 둘 다 어쨌든 Exception의 상속을 받아서 같이 처리됨. 그럼 둘 다 동일한 예외 메시지를 날리면 잘못 된 거 아닌가?</p>
<blockquote>
<p>밤에 이런 생각이 들어서 챗지피티에게 물어봤습니다요 ㅋ
낼 튜터님께 여쭤봐야지 ㅋ</p>
</blockquote>
<br />

<p>아주 중요한 본질을 짚었어. 이게 바로 “예외 정보의 정확성 유지”와 “catch 범위 관리” 사이의 <code>trade-off</code> 문제야. 하나씩 풀어줄게.</p>
<p>🔥 문제 요약</p>
<p><code>상황</code>
catch (Exception e)로 모든 예외를 통합적으로 처리하려고 했는데
실제 예외는 다양한 위치에서 다르게 발생할 수 있음 (예: User 없음, Todo 없음, DB 오류 등)
그런데 catch 안에서 던지는 예외는 단일 의미로 고정됨</p>
<p>예: throw new ManagerCreateFailedException(...)</p>
<p>👉 결국 실제 오류 원인과 다른 예외 메시지를 사용자에게 전달할 위험이 생김</p>
<blockquote>
<p>이러고 지피티가 다양한 방법을 추천해줬는데 결국 내용은
니가 에러날 상황을 다 파악하고 에러별 catch문을 명확하게 일일이 잡아주라는 거였음^^</p>
</blockquote>
<pre><code class="language-java">// 예외 터지면 그대로 던져도 되는 부분
User user = userRepository.findById(...).orElseThrow(() -&gt; new InvalidRequestException(...));
Todo todo = ... // 동일

try {
    Manager saved = managerRepository.save(new Manager(...));
    logService.saveLog(saved, false);
    return new ManagerSaveResponse(...);
} catch (DataAccessException e) {
    logService.saveLog(newManagerUser, true);
    throw new ManagerCreateFailedException(&quot;매니저 저장 실패&quot;, e);
}</code></pre>
<br />

<blockquote>
<p><a href="https://minjooig.tistory.com/142">https://minjooig.tistory.com/142</a>
참고 블로그</p>
</blockquote>